File: mobile/lib/features/lab/data/note_attachment.dart
/// Modelos de `/api/v1/notes/{id}/attachments` — capturas de gráficos y su capa de dibujo.
///
/// Espejan `app/schemas/note_attachment.py`. Las dos reglas del contrato, que acá se hacen cumplir
/// por construcción:
///
///   1. **Las coordenadas son NORMALIZADAS (0..1), nunca píxeles.** [NormalizedPoint] recorta en el
///      constructor, así que es imposible construir un punto fuera de rango — que el backend
///      rechazaría con 422 después de que el usuario ya dibujó el trazo. La conversión a píxeles
///      vive en un solo lugar ([DrawingGeometry]) y depende del rectángulo REAL de la imagen, no del
///      tamaño del widget.
///   2. **Cada herramienta tiene su aridad.** Una recta con un punto o un texto vacío son un 422 del
///      backend; [DrawingShape.isDrawable] deja detectarlo antes de mandar y de dibujar.
library;

import 'dart:ui' show Color;

import 'package:flutter/foundation.dart';

/// Herramientas del canvas. Los nombres del wire son los de `ShapeKind` en el backend.
enum ShapeKind { line, arrow, rect, ellipse, freehand, text }

extension ShapeKindWire on ShapeKind {
  String get wireValue => switch (this) {
        ShapeKind.line => 'LINE',
        ShapeKind.arrow => 'ARROW',
        ShapeKind.rect => 'RECT',
        ShapeKind.ellipse => 'ELLIPSE',
        ShapeKind.freehand => 'FREEHAND',
        ShapeKind.text => 'TEXT',
      };

  String get displayName => switch (this) {
        ShapeKind.line => 'Línea',
        ShapeKind.arrow => 'Flecha',
        ShapeKind.rect => 'Rectángulo',
        ShapeKind.ellipse => 'Elipse',
        ShapeKind.freehand => 'Trazo libre',
        ShapeKind.text => 'Texto',
      };

  /// Cuántos puntos necesita la forma. `null` = variable (el trazo libre).
  ///
  /// Es la misma tabla que valida el backend. Se replica acá para poder rechazar un trazo imposible
  /// ANTES de mandarlo: enterarse por un 422 después de dibujar es enterarse tarde.
  int? get requiredPoints => switch (this) {
        ShapeKind.line || ShapeKind.arrow => 2,
        ShapeKind.rect || ShapeKind.ellipse => 2,
        ShapeKind.text => 1,
        ShapeKind.freehand => null,
      };
}

/// `line` como fallback: es la herramienta más simple, así que una forma de un tipo que este cliente
/// no conoce se dibuja como una recta en vez de romper el canvas entero.
ShapeKind shapeKindFromWire(String? value) => switch (value) {
      'ARROW' => ShapeKind.arrow,
      'RECT' => ShapeKind.rect,
      'ELLIPSE' => ShapeKind.ellipse,
      'FREEHAND' => ShapeKind.freehand,
      'TEXT' => ShapeKind.text,
      _ => ShapeKind.line,
    };

/// Un punto en coordenadas relativas a la imagen: (0,0) arriba a la izquierda, (1,1) abajo a la
/// derecha.
///
/// **Se recorta en el constructor.** Un dedo que se va del borde de la imagen mientras dibuja
/// produciría un 1.03 que el backend rechaza con 422 — y el usuario perdería el trazo entero por
/// haberse pasado dos píxeles. Recortar conserva la intención y respeta el contrato.
@immutable
class NormalizedPoint {
  NormalizedPoint(double x, double y)
      : x = x.isFinite ? x.clamp(0.0, 1.0) : 0.0,
        y = y.isFinite ? y.clamp(0.0, 1.0) : 0.0;

  factory NormalizedPoint.fromJson(Map<String, dynamic> json) => NormalizedPoint(
        (json['x'] as num?)?.toDouble() ?? 0,
        (json['y'] as num?)?.toDouble() ?? 0,
      );

  final double x;
  final double y;

  Map<String, dynamic> toJson() => {'x': x, 'y': y};

  @override
  bool operator ==(Object other) =>
      other is NormalizedPoint && other.x == x && other.y == y;

  @override
  int get hashCode => Object.hash(x, y);

  @override
  String toString() =>
      'NormalizedPoint(${x.toStringAsFixed(3)}, ${y.toStringAsFixed(3)})';
}

/// Convierte un `Color` al `#aarrggbb` que espera el backend.
///
/// Se emite SIEMPRE con alfa, aunque sea opaco: un resaltado semitransparente y un trazo sólido son
/// dos herramientas distintas del editor, y perder el canal al guardar convertiría uno en el otro.
String colorToHex(Color color) =>
    '#${color.toARGB32().toRadixString(16).padLeft(8, '0')}';

/// Parsea `#rrggbb` o `#aarrggbb`. Devuelve `null` si no es ninguno de los dos.
Color? hexToColor(String? value) {
  if (value == null) return null;
  final hex = value.startsWith('#') ? value.substring(1) : value;
  if (hex.length != 6 && hex.length != 8) return null;
  final parsed = int.tryParse(hex, radix: 16);
  if (parsed == null) return null;
  // Un `#rrggbb` sin alfa es opaco: es la convención de CSS y la que usa el resto del contrato.
  return Color(hex.length == 6 ? 0xFF000000 | parsed : parsed);
}

@immutable
class DrawingShape {
  const DrawingShape({
    required this.id,
    required this.kind,
    required this.points,
    required this.color,
    this.strokeWidth = 2.0,
    this.text,
  });

  factory DrawingShape.fromJson(Map<String, dynamic> json) => DrawingShape(
        id: json['id'] as String,
        kind: shapeKindFromWire(json['kind'] as String?),
        points: [
          for (final point in (json['points'] as List?) ?? [])
            NormalizedPoint.fromJson(point as Map<String, dynamic>),
        ],
        // Un color ilegible cae al cian de acento en vez de romper el render: el trazo se ve, aunque
        // con otro tono, y el usuario puede corregirlo.
        color: hexToColor(json['color'] as String?) ?? const Color(0xFF22D3EE),
        strokeWidth: (json['stroke_width'] as num?)?.toDouble() ?? 2.0,
        text: json['text'] as String?,
      );

  final String id;
  final ShapeKind kind;
  final List<NormalizedPoint> points;
  final Color color;
  final double strokeWidth;

  /// Solo para [ShapeKind.text].
  final String? text;

  /// ¿La forma se puede dibujar y guardar?
  ///
  /// Es la misma regla que valida el backend. Se chequea acá para no mandar una forma que va a
  /// volver como 422 y, sobre todo, para no intentar pintar una recta sin segundo punto — que sería
  /// una excepción en medio del `paint`, es decir un frame rojo.
  bool get isDrawable {
    final required = kind.requiredPoints;
    if (required != null && points.length != required) return false;
    if (kind == ShapeKind.freehand && points.length < 2) return false;
    if (kind == ShapeKind.text) return (text ?? '').trim().isNotEmpty;
    return true;
  }

  DrawingShape copyWith({
    List<NormalizedPoint>? points,
    Color? color,
    double? strokeWidth,
    String? text,
  }) =>
      DrawingShape(
        id: id,
        kind: kind,
        points: points ?? this.points,
        color: color ?? this.color,
        strokeWidth: strokeWidth ?? this.strokeWidth,
        text: text ?? this.text,
      );

  Map<String, dynamic> toJson() => {
        'id': id,
        'kind': kind.wireValue,
        'points': [for (final point in points) point.toJson()],
        'color': colorToHex(color),
        'stroke_width': strokeWidth,
        // El backend RECHAZA `text` en una forma que no es TEXT, así que solo se emite donde
        // corresponde: mandarlo siempre sería un 422 en cada línea que se dibuje.
        if (kind == ShapeKind.text) 'text': text,
      };

  @override
  bool operator ==(Object other) =>
      other is DrawingShape &&
      other.id == id &&
      other.kind == kind &&
      listEquals(other.points, points) &&
      other.color == color &&
      other.strokeWidth == strokeWidth &&
      other.text == text;

  @override
  int get hashCode =>
      Object.hash(id, kind, Object.hashAll(points), color, strokeWidth, text);
}

@immutable
class DrawingLayer {
  const DrawingLayer({this.version = 1, this.shapes = const []});

  factory DrawingLayer.fromJson(Map<String, dynamic> json) => DrawingLayer(
        version: json['version'] as int? ?? 1,
        shapes: [
          for (final shape in (json['shapes'] as List?) ?? [])
            DrawingShape.fromJson(shape as Map<String, dynamic>),
        ],
      );

  static const empty = DrawingLayer();

  final int version;
  final List<DrawingShape> shapes;

  bool get isEmpty => shapes.isEmpty;
  bool get isNotEmpty => shapes.isNotEmpty;

  /// Solo las formas que se pueden pintar.
  ///
  /// Una capa guardada por una versión anterior puede traer una forma que este cliente no sabe
  /// dibujar; se salta esa y se pintan las demás, en vez de dejar la captura sin anotaciones.
  List<DrawingShape> get drawable =>
      shapes.where((shape) => shape.isDrawable).toList();

  DrawingLayer copyWith({List<DrawingShape>? shapes}) =>
      DrawingLayer(version: version, shapes: shapes ?? this.shapes);

  Map<String, dynamic> toJson() => {
        'version': version,
        'shapes': [for (final shape in shapes) shape.toJson()],
      };

  @override
  bool operator ==(Object other) =>
      other is DrawingLayer &&
      other.version == version &&
      listEquals(other.shapes, shapes);

  @override
  int get hashCode => Object.hash(version, Object.hashAll(shapes));
}

@immutable
class NoteAttachment {
  const NoteAttachment({
    required this.id,
    required this.noteId,
    required this.ticker,
    required this.contentType,
    required this.byteSize,
    required this.width,
    required this.height,
    required this.caption,
    required this.source,
    required this.drawing,
    required this.imageUrl,
    required this.createdAt,
    required this.updatedAt,
  });

  factory NoteAttachment.fromJson(Map<String, dynamic> json) => NoteAttachment(
        id: json['id'] as String,
        noteId: json['note_id'] as String,
        ticker: json['ticker'] as String?,
        contentType: json['content_type'] as String? ?? 'image/png',
        byteSize: json['byte_size'] as int? ?? 0,
        width: json['width'] as int?,
        height: json['height'] as int?,
        caption: json['caption'] as String?,
        source: json['source'] as String?,
        drawing: DrawingLayer.fromJson(
          (json['drawing'] as Map<String, dynamic>?) ?? const {},
        ),
        imageUrl: json['image_url'] as String,
        createdAt: DateTime.parse(json['created_at'] as String),
        updatedAt: DateTime.parse(json['updated_at'] as String),
      );

  final String id;
  final String noteId;
  final String? ticker;
  final String contentType;
  final int byteSize;
  final int? width;
  final int? height;
  final String? caption;
  final String? source;
  final DrawingLayer drawing;

  /// Ruta relativa de los bytes, tal como la arma el backend. NO se concatena del lado del cliente:
  /// si la ruta cambia, esta app sigue funcionando.
  final String imageUrl;

  final DateTime createdAt;
  final DateTime updatedAt;

  /// Relación ancho/alto de la imagen, si el backend la conoce.
  ///
  /// `null` cuando el cliente que subió la captura no declaró las dimensiones. En ese caso el visor
  /// tiene que sacarla de la imagen ya decodificada — dibujar las anotaciones antes de saberla las
  /// pondría en el lugar equivocado.
  double? get aspectRatio {
    final w = width;
    final h = height;
    if (w == null || h == null || w <= 0 || h <= 0) return null;
    return w / h;
  }

  bool get hasAnnotations => drawing.isNotEmpty;

  int get shapeCount => drawing.shapes.length;

  NoteAttachment copyWith({DrawingLayer? drawing, String? caption}) =>
      NoteAttachment(
        id: id,
        noteId: noteId,
        ticker: ticker,
        contentType: contentType,
        byteSize: byteSize,
        width: width,
        height: height,
        caption: caption ?? this.caption,
        source: source,
        drawing: drawing ?? this.drawing,
        imageUrl: imageUrl,
        createdAt: createdAt,
        updatedAt: updatedAt,
      );
}

/// Lo que se manda al subir una captura.
@immutable
class AttachmentDraft {
  const AttachmentDraft({
    required this.imageBytes,
    this.contentType = 'image/png',
    this.ticker,
    this.caption,
    this.source,
    this.width,
    this.height,
    this.drawing = DrawingLayer.empty,
  });

  final Uint8List imageBytes;
  final String contentType;
  final String? ticker;
  final String? caption;
  final String? source;
  final int? width;
  final int? height;
  final DrawingLayer drawing;
}

File: app/schemas/note_attachment.py
"""Schemas de `/api/v1/notes/{id}/attachments` — capturas de gráficos y su capa de dibujo.

Dos decisiones de contrato que sostienen todo el módulo:

  1. **Las coordenadas del dibujo son NORMALIZADAS (0..1), no píxeles.** La misma captura se muestra
     a 320px en un teléfono y a 900px en el panel del escritorio; con coordenadas absolutas, una
     flecha que apunta a un máximo de precio apuntaría al aire en la otra pantalla. Normalizar deja
     el trazo pegado al punto del gráfico que el usuario marcó, en cualquier tamaño.
  2. **La forma se valida por tipo.** Una recta con tres puntos, o un texto sin texto, son datos que
     ningún cliente puede dibujar: aceptarlos guardaría basura que rompe el canvas al releerla, y el
     error aparecería a la semana siguiente, lejos de donde se originó.

La imagen viaja en base64 porque el cuerpo es JSON —el dibujo tiene que llegar en el mismo request
que la captura, y un `multipart` con un campo JSON adentro es incómodo desde el cliente—, pero se
**guarda en crudo**: conservar el base64 desperdiciaría un 33% del espacio para siempre.
"""

from __future__ import annotations

import base64
import binascii
from datetime import datetime
from enum import Enum
from uuid import UUID

from pydantic import (
    BaseModel,
    ConfigDict,
    Field,
    PrivateAttr,
    field_validator,
    model_validator,
)

# Tope de la imagen decodificada. Una captura de un chart ronda los 200-600 KB; 4 MB deja lugar de
# sobra para una pantalla retina completa y corta antes de que una subida por error llene la tabla.
MAX_IMAGE_BYTES = 4 * 1024 * 1024

# El base64 infla ~33%, más el margen del `data:` URI que algunos clientes anteponen.
_MAX_ENCODED_LENGTH = (MAX_IMAGE_BYTES * 4) // 3 + 1024

# Tope de formas por capa. Un dibujo a mano alzada largo puede tener cientos de puntos pero no miles
# de formas; sin tope, un bug del cliente podría empujar un JSON de decenas de MB.
MAX_SHAPES = 500

# Tope de puntos de un trazo libre. Es lo que acota el tamaño de la capa: 2000 puntos son un garabato
# muy largo, y el cliente puede simplificar antes de mandar.
MAX_POINTS = 2000

# Formatos aceptados. Lista blanca y no "cualquier image/*": el endpoint que devuelve los bytes
# reenvía este valor como `Content-Type`, y aceptar `image/svg+xml` convertiría el adjunto en un
# vector de scripting sobre el dominio de la app.
ALLOWED_CONTENT_TYPES = frozenset({"image/png", "image/jpeg", "image/webp"})


def _matches_signature(raw: bytes, content_type: str) -> bool:
    """¿Los bytes son realmente del formato declarado?

    Se chequea la firma y no se confía en el `content_type` del cliente porque el desajuste es
    permanente y silencioso: un adjunto que dice `image/png` y no lo es se guarda sin error y se
    muestra como un ícono roto para siempre, con el fallo apareciendo semanas después de la subida y
    lejos de donde se originó. Además el endpoint de bytes reenvía ese header, así que dejar que la
    declaración y el contenido difieran habilita que el navegador olfatee algo distinto de una imagen.
    """

    match content_type:
        case "image/png":
            return raw.startswith(b"\x89PNG\r\n\x1a\n")
        case "image/jpeg":
            return raw.startswith(b"\xff\xd8\xff")
        case "image/webp":
            # RIFF <4 bytes de tamaño> WEBP
            return len(raw) >= 12 and raw[:4] == b"RIFF" and raw[8:12] == b"WEBP"
        case _:  # pragma: no cover - la lista blanca ya lo filtró
            return False


class ShapeKind(str, Enum):
    """Herramientas del canvas. Cada una define cuántos puntos necesita, y esa es la validación."""

    LINE = "LINE"
    ARROW = "ARROW"
    RECT = "RECT"
    ELLIPSE = "ELLIPSE"
    FREEHAND = "FREEHAND"
    TEXT = "TEXT"


class DrawingPoint(BaseModel):
    """Un punto en coordenadas normalizadas: (0,0) es la esquina superior izquierda de la imagen y
    (1,1) la inferior derecha.
    """

    # `allow_inf_nan=False` es redundante HOY —los límites `ge`/`le` ya rechazan `NaN` e `inf`— y
    # está igual porque expresa la intención por separado del rango: si mañana alguien permite
    # coordenadas fuera de 0..1 para una herramienta nueva, un `NaN` seguiría rechazado en vez de
    # colarse y dejar el canvas del cliente sin dibujar la capa entera.
    model_config = ConfigDict(extra="forbid", allow_inf_nan=False)

    x: float = Field(ge=0, le=1)
    y: float = Field(ge=0, le=1)


class DrawingShape(BaseModel):
    model_config = ConfigDict(extra="forbid")

    # Id del lado del cliente: es lo que le permite editar o borrar una forma puntual sin reenviar la
    # capa entera desde cero y sin que el servidor tenga que devolverle ids nuevos.
    id: str = Field(min_length=1, max_length=64)

    kind: ShapeKind
    points: list[DrawingPoint] = Field(min_length=1, max_length=MAX_POINTS)

    # `#rrggbb` o `#aarrggbb`. Se valida el formato porque este valor termina en un parser de color
    # del cliente, donde una cadena arbitraria es una excepción en tiempo de render.
    color: str = Field(pattern=r"^#(?:[0-9a-fA-F]{6}|[0-9a-fA-F]{8})$")

    stroke_width: float = Field(default=2.0, gt=0, le=64)

    # Solo para `TEXT`. Se acota porque una anotación sobre un gráfico es una etiqueta, no un párrafo.
    text: str | None = Field(default=None, max_length=280)

    @model_validator(mode="after")
    def _check_arity(self) -> DrawingShape:
        """Cada herramienta necesita una cantidad concreta de puntos.

        Se valida acá y no en el cliente porque es la regla que decide si la forma se puede dibujar:
        una `LINE` con un solo punto no tiene dirección, y un `RECT` con tres no tiene esquinas. Una
        capa con una forma así se guarda sin error y explota al releerla.
        """

        count = len(self.points)
        expected = {
            ShapeKind.LINE: 2,
            ShapeKind.ARROW: 2,
            ShapeKind.RECT: 2,
            ShapeKind.ELLIPSE: 2,
            ShapeKind.TEXT: 1,
        }.get(self.kind)

        if expected is not None and count != expected:
            raise ValueError(
                f"Una forma {self.kind.value} necesita exactamente {expected} "
                f"{'punto' if expected == 1 else 'puntos'}, llegaron {count}."
            )
        if self.kind is ShapeKind.FREEHAND and count < 2:
            raise ValueError("Un trazo libre necesita al menos 2 puntos.")

        if self.kind is ShapeKind.TEXT:
            if self.text is None or not self.text.strip():
                raise ValueError("Una anotación de texto no puede estar vacía.")
        elif self.text is not None:
            # Se rechaza en vez de ignorarlo: un texto que el cliente cree haber guardado y que nadie
            # va a dibujar es peor que un 422.
            raise ValueError(
                f"El campo `text` solo aplica a las formas TEXT, no a {self.kind.value}."
            )

        return self


class DrawingLayer(BaseModel):
    """La capa vectorial completa de un adjunto.

    Se manda y se guarda SIEMPRE entera: el canvas conoce su estado completo, y un protocolo de
    parches por forma agregaría resolución de conflictos para un recurso que edita un solo usuario en
    una sola pantalla.
    """

    model_config = ConfigDict(extra="forbid")

    # Versión del formato. Existe desde el día uno para que, cuando aparezca una herramienta nueva, un
    # cliente viejo pueda decir "esta capa la dibujó una versión que no entiendo" en vez de dibujar
    # una parte y callarse el resto.
    version: int = Field(default=1, ge=1, le=100)

    shapes: list[DrawingShape] = Field(default_factory=list, max_length=MAX_SHAPES)

    @property
    def is_empty(self) -> bool:
        return not self.shapes


class NoteAttachmentCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    # Base64 de la imagen. Se acepta con o sin el prefijo `data:image/png;base64,` que anteponen
    # `canvas.toDataURL()` y varios pickers: obligar al cliente a pelarlo sería un 422 por una
    # diferencia que el servidor puede resolver solo.
    image_data: str = Field(min_length=4, max_length=_MAX_ENCODED_LENGTH)

    content_type: str = Field(default="image/png", max_length=64)
    ticker: str | None = Field(default=None, min_length=1, max_length=20)
    caption: str | None = Field(default=None, max_length=300)
    source: str | None = Field(default=None, max_length=64)
    width: int | None = Field(default=None, gt=0, le=20000)
    height: int | None = Field(default=None, gt=0, le=20000)

    # Un adjunto puede subirse sin dibujo: primero se pega la captura y después se anota.
    drawing: DrawingLayer = Field(default_factory=DrawingLayer)

    @field_validator("content_type")
    @classmethod
    def _known_type(cls, value: str) -> str:
        normalized = value.strip().lower()
        if normalized not in ALLOWED_CONTENT_TYPES:
            raise ValueError(
                "Formato no soportado. Se aceptan "
                f"{', '.join(sorted(ALLOWED_CONTENT_TYPES))}."
            )
        return normalized

    @field_validator("ticker")
    @classmethod
    def _upper(cls, value: str | None) -> str | None:
        return value.strip().upper() if value else None

    # Los bytes ya decodificados. Privado porque el que vale es el resultado del validador, no una
    # decodificación que cada llamador pueda repetir con otros criterios.
    _decoded: bytes = PrivateAttr(default=b"")

    @model_validator(mode="after")
    def _decode(self) -> NoteAttachmentCreate:
        """Decodifica y verifica la imagen DURANTE la validación del request.

        Va acá y no en el servicio para que un base64 corrupto sea un 422 —que es exactamente lo que
        es— en vez de una excepción a mitad de la escritura, que llegaría al cliente como un 500 sin
        decirle qué mandó mal. Los validadores de campo ya corrieron, así que `content_type` está
        normalizado y pasó por la lista blanca.
        """

        self._decoded = self._decode_image_data()
        return self

    @property
    def image(self) -> bytes:
        """Los bytes crudos de la imagen, listos para guardar."""

        return self._decoded

    def _decode_image_data(self) -> bytes:
        payload = self.image_data.strip()
        # `data:image/png;base64,iVBOR...` — se descarta el encabezado y se usa el `content_type`
        # declarado, que ya pasó por la lista blanca. Confiar en el del URI dejaría entrar un tipo
        # que el validador rechazó.
        if payload.startswith("data:"):
            _, _, payload = payload.partition(",")

        try:
            raw = base64.b64decode(payload, validate=True)
        except (binascii.Error, ValueError) as exc:
            raise ValueError("La imagen no es base64 válido.") from exc

        if not raw:
            raise ValueError("La imagen está vacía.")
        if len(raw) > MAX_IMAGE_BYTES:
            raise ValueError(
                f"La imagen supera el máximo de {MAX_IMAGE_BYTES // (1024 * 1024)} MB."
            )
        if not _matches_signature(raw, self.content_type):
            raise ValueError(
                f"Los bytes no corresponden a un {self.content_type}. "
                "Revisá el formato declarado."
            )
        return raw


class NoteAttachmentUpdate(BaseModel):
    """PUT de la capa de dibujo.

    Es un PUT y no un PATCH porque el canvas manda su estado completo: mandar media capa y esperar
    que el servidor la fusione con la anterior dejaría las formas borradas vivas para siempre.

    Los metadatos (`caption`, `ticker`) sí son opcionales y se comportan como un PATCH; para
    desvincular el símbolo hay que mandar `null` explícito, igual que en el resto del Lab.
    """

    model_config = ConfigDict(extra="forbid")

    drawing: DrawingLayer
    caption: str | None = Field(default=None, max_length=300)
    ticker: str | None = Field(default=None, min_length=1, max_length=20)

    @field_validator("ticker")
    @classmethod
    def _upper(cls, value: str | None) -> str | None:
        return value.strip().upper() if value else None


class NoteAttachmentRead(BaseModel):
    """Un adjunto SIN sus bytes.

    La imagen se pide aparte (`GET .../{id}/image`) y por eso este schema no la incluye: una nota con
    diez capturas devolvería decenas de megabytes en base64 cada vez que se abre, casi todos para
    miniaturas que el navegador ya tiene cacheadas.
    """

    model_config = ConfigDict(strict=True, extra="forbid")

    id: UUID
    note_id: UUID
    ticker: str | None
    content_type: str
    byte_size: int = Field(ge=0)
    width: int | None
    height: int | None
    caption: str | None
    source: str | None

    # La capa vectorial sí viaja entera: son cientos de bytes y el canvas la necesita para dibujar
    # las anotaciones encima de la imagen apenas ésta carga.
    drawing: DrawingLayer

    # URL relativa de los bytes. La arma el backend en vez de dejar que el cliente la concatene: si
    # la ruta cambia, los clientes viejos siguen funcionando.
    image_url: str

    created_at: datetime
    updated_at: datetime

    @property
    def shape_count(self) -> int:
        return len(self.drawing.shapes)
